internal/cortex/frontend/transport/handler.go

```diff
@@ -122,14 +153,54 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
 	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
 
+	if f.lru != nil {
+		// Store query expression
+		queryExpressionNormalized = f.regex.ReplaceAllString(r.URL.Query().Get("query"), " ")
+
+		// Time range length for queries, if either of "start" or "end" are not present, return 0
+		getQueryRangeSeconds := func() int {
+			start, err := strconv.Atoi(r.URL.Query().Get("start"))
+			if err != nil {
+				return 0
+			}
+			end, err := strconv.Atoi(r.URL.Query().Get("end"))
+			if err != nil {
+				return 0
+			}
+			return end - start
+		}
+
+		//Store query time range length
+		queryExpressionRangeLength = getQueryRangeSeconds()
+
+		// Check if query in cache and whether value exceeds time range length
+		if value, ok := f.lru.Get(queryExpressionNormalized); ok && value.(int) >= queryExpressionRangeLength {
+			w.WriteHeader(http.StatusForbidden)
+			level.Info(util_log.WithContext(r.Context(), f.log)).Log("msg", "found query in cache, caused error", "filter for this error log with following query expression ", queryExpressionNormalized)
+			return
+		}
+	}
+
 	startTime := time.Now()
 	resp, err := f.roundTripper.RoundTrip(r)
 	queryResponseTime := time.Since(startTime)
 
 	if err != nil {
 		writeError(w, err)
+		queryString = f.parseRequestQueryString(r, buf)
+
+		if f.lru != nil {
+			// If error should be cached, store it in cache
+			if CacheableError(resp.StatusCode) {
+				f.lru.Add(queryExpressionNormalized, queryExpressionRangeLength)
+
+				level.Info(util_log.WithContext(r.Context(), f.log)).Log(
```

Comment:
ditto.

Reply:
is it okay to leave it for now while I am testing and remove after?

Reply:
Cached/blocked queries should be rare and it's an important event for debugging. We should keep this log line in case if any user questions us why a query is blocked. We need this log line to get the clear answer. A metric won't help in this case.